rs-matter/src/transport/network/btp/session.rs

```diff
@@ -0,0 +1,636 @@
+    rem_msg_len: u16,
+}
+
+impl RecvWindow {
+    /// Initialize a new receiving window with the provided window size.
+    #[inline(always)]
+    pub const fn new(window_size: u8) -> Self {
+        Self {
+            buf: RingBuf::new(),
+            buf_messages_ct: 0,
+            level: window_size,
+            ack_level: 0,
+            ack_seq: 255,
+            received_at: Instant::MAX,
+            rem_msg_len: 0,
+        }
+    }
+
+    /// Process an incoming BTP segment, updating the state of the window accordingly.
+    fn accept_incoming(&mut self, hdr: &BtpHdr, payload: &[u8], mtu: u16) -> Result<(), Error> {
+        // Check received packet integrity, as per the Matter Core spec
+        self.check_data_integrity(hdr, payload, mtu)?;
+
+        if let Some(msg_len) = hdr.get_msg_len() {
+            if msg_len <= mtu && !hdr.is_final() {
+                warn!("RX data integrity failure: An SDU that fits in a single BTP segment must be final");
+                Err(ErrorCode::InvalidData)?;
+            }
+
+            self.rem_msg_len = msg_len;
+
+            if msg_len > 0 {
+                if self.buf.free() >= core::mem::size_of::<u16>() {
+                    // New SDU; skip 0-length ones as they do not contain Matter messages
+                    self.buf.push(&u16::to_le_bytes(msg_len));
+                } else {
+                    Err(ErrorCode::NoSpace)?;
+                }
+            }
+        }
+
+        if self.rem_msg_len < payload.len() as u16 {
+            warn!("RX data integrity failure: Packet contains more data than the message length");
+            Err(ErrorCode::InvalidData)?;
+        } else {
+            self.rem_msg_len -= payload.len() as u16;
+            if hdr.is_final() && self.rem_msg_len > 0 {
+                warn!("RX data integrity failure: Packet is final but the message length is not reached");
+                Err(ErrorCode::InvalidData)?;
+            }
+        }
+
+        if self.buf.free() >= payload.len() {
+            self.buf.push(payload);
+        } else {
+            Err(ErrorCode::NoSpace)?;
+        }
+
+        self.level -= 1;
+        // Unwrap is safe because we are only processing BTP data segments here and they always have a sequence number
+        self.ack_seq = hdr.get_seq().unwrap();
+        self.ack_level += 1;
+        self.received_at = Instant::now();
+
+        if hdr.is_final() && !payload.is_empty() {
+            self.buf_messages_ct += 1;
+        }
+
+        Ok(())
+    }
+
+    fn check_data_integrity(&self, hdr: &BtpHdr, payload: &[u8], mtu: u16) -> Result<(), Error> {
+        if hdr.is_handshake() // Handshake packets are not allowed here
+            || hdr.get_opcode().is_some()
+        {
+            warn!("RX data integrity failure: Data and standalone ACK packets must not have an opcode");
+            return Err(ErrorCode::InvalidData.into());
```

Comment:
I think you might have a point here. 

Given that we hard-code the ring-buffer size to be `MAX_MESSAGE_SIZE = MAX_RX_PACKET_SIZE * 2` we should _never, ever_ get in a situation where the ring-buffer is full, unless the peer is mis-behaving and sending us stuff even when our recv window is full and we did not send an ACK to let the other party know it can continue sending.

I'll change to `ErrorCode::InvalidData` the existing two `ErrorCode::NoSpace` error reports.